java-dynamic-sqs-listener-api/src/main/java/com/jashmore/sqs/broker/MessageBroker.java

```diff
@@ -0,0 +1,55 @@
+package com.jashmore.sqs.broker;
+
+import com.jashmore.sqs.processor.MessageProcessor;
+import com.jashmore.sqs.retriever.MessageRetriever;
+
+import java.util.concurrent.Future;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Broker used to co-ordinate the retrieval and processing of messages from a remote queue.
+ *
+ * <p>If you were to consider this library as similar to a pub-sub system, this would be considered the broker that transports messages from the publisher
+ * to the subscriber.  It will request for messages from the {@link MessageRetriever} and delegate the processing to a corresponding {@link MessageProcessor}.
+ *
+ * <p>As this is a non-blocking class that will start the message process in a background thread there is the possibility of usage of this class in multiple
+ * threads. Therefore, implementations of this class must be thread safe.
+ */
+@ThreadSafe
+public interface MessageBroker {
+    /**
+     * Start the brokerage of messages from the queue to the processor.
+     *
+     * <p>This must be non-blocking and start the process on a background thread.
+     *
+     * <p>If this broker has already been started then any subsequent calls to this method will do no action until {@link #stop(boolean)} has been
```

Comment:
What exactly does "will do no action until ..." mean? Will it wait? Or will it just short-circuit. Either case, I am torn about whether this should be an exception case or at least give the caller a way to tell that their call to start was ignored because it was already started. This could be an error case that we want to have the ability to signal. Maybe not an exception, but a means to see if any action occurred - status object of the broker or a boolean?

Reply:
Yeah I have been jumping around on this detail. Originally I had it throw an exception and then before I just had it do nothing (or in the stop case, just return a resolved promise).  I don't think there would be any case where we would call start twice and it would indicate a bug so maybe an exception is acceptable?

Reply:
That seems the best bet - throw exception if start called twice. You could simply re-use `java.lang.IllegalStateException`